### PR TITLE
Adding Knowledge Distillation (Image Classification)

### PR DESCRIPTION
**Pull request by asusevski**

### CODE REVIEW

File: chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb [hunk 1]
```diff
@@ -0,0 +1,536 @@
+{
```
Comment (merveenoyan) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Maybe you can give a link to the DistilBERT and DistilGPT? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (merveenoyan) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>You can have this before initializing model right before image processing class</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (merveenoyan) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>You can have comments in markdown section and not the code </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (merveenoyan) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Maybe we could also push the model to Hub and give a link to people can use it off the shelf :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

**Review by merveenoyan — COMMENTED** [review]
I mostly did nitpicking, thanks a lot for this notebook!

File: chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb (hunk 1 above)
Comment (MKhalusova) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>1) We'll begin by importing the required <s>datasets</s> libraries:</p><p>2) "Presumably, we've all had teachers who teach by having us memorize correct answers, analogous to Supervised Training of machine learning models where we provide a labelled dataset to train on. Instead of having a model train on labels, however, we can pursue <a href="https://arxiv.org/abs/1503.02531" target="_blank">Knowledge Distillation</a> as an alternative to arrive at a much smaller model that can perform comparably to the larger model, and much faster to boot. "</p><p>re:2 I don't fully understand what you're saying with this metaphor. "by having us memorize correct answers" sounds like overfitting, rather than supervised training. </p><p>"Instead of having a model train on labels" is somewhat misleading, as the teacher model is fine-tuned on labeled data. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (MKhalusova) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Why is there a need to create an Image Processor? Why is ViTImageProcessor not enough?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Will add!! Thanks for the callout!</p>

[the remark this answers is not in the document]
Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Will take into consideration!</p>

[the remark this answers is not in the document]
Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Thanks, will do!</p>

Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Ok! The problem is the results haven't been that great starting with a randomized ViT -- so far been unable to attain higher than 35% accuracy on test set, trained for about 30 epochs. I'm still trying to improve performance, I will push once performance improves</p>

Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>1) Thanks for the typo catch!</p><p> </p><p>2) Fair points, I will re-word and update!!</p>

Comment (asusevski) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Good question, I am actually curious if you or anyone else had any insight into this -- I noticed significantly poorer performance when training with ViTImageProcessor, and I'm not entirely sure why. To dig deeper, I used ViTImageProcessor on some examples and found that they ended up looking different than they ought to (<a href="https://imgur.com/a/4nTtW7L" target="_blank">https://imgur.com/a/4nTtW7L</a>). So I decided to use write my own Processor for now. Might you have any insight into what is going on?</p>

Comment (MKhalusova) [reply chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: I think @merveenoyan or @NielsRogge can answer this question better than I could.

File: chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
```
Comment (merveenoyan) [inline chapters/en/Unit 3 - Vision Transformers/KnowledgeDistillation.ipynb:1]: <p>Lastly, can you add push_to_hub here so people can try your model rightaway? You can also give the link to the model repository.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/111/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

**Review by johko — APPROVED** [review]
Very nice notebooks, thanks a lot for creating these :pray: 

After you handle the things mentioned by Merve above, I think we're ready to merge :rocket:

**Comment by asusevski** [comment]
Sounds good, I'm currently away for Christmas but I'll address these minor changes by the 27th!

**Comment by asusevski** [comment]
@merveenoyan @johko I know I'm fairly late, I only recently returned from spending time away for the holidays, I was hoping you could check now if all the changes are good to merge?

**Review by merveenoyan — APPROVED** [review]
@asusevski looks good to me! on the other file you're contributing, it's already merged to the main repository. maybe you can sync with upstream (or this repository) or remove that file altogether and we can merge

**Comment by asusevski** [comment]
done! I thought i had removed it already for some reason. Sorry about that!!
